cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -1828,6 +1828,148 @@ void AddUtf8Length(FunctionRegistry* registry) {
   DCHECK_OK(registry->AddFunction(std::move(func)));
 }
 
+// binary join
+
+template <typename Type>
+struct Join {
+  using ArrayType = typename TypeTraits<Type>::ArrayType;
+  using ListArrayType = ListArray;
+  using offset_type = typename Type::offset_type;
+  using BuilderType = typename TypeTraits<Type>::BuilderType;
+
+  static void Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    if (batch[0].kind() == Datum::SCALAR) {
```

Comment:
For readability, can you split this function into different functions based on the various input kinds, e.g. `ExecArrayArray`, `ExecArrayScalar`, `ExecScalarArray`, `ExecScalarScalar`?